RationaleMCP/0027/HindleyMilner.md

```diff
@@ -0,0 +1,210 @@
+* If `var` doesn't have a declared unit attribute, and has constant component variablity, `e.unit` is equal to `empty` unit.
+* Otherwise, `e.unit` is equal to `var.unit`.
+
+If `var` has a declared unit attribute, `var.unit` must be compatible with it.
+
+### Literals
+Literals have `empty` unit.
+
+### Binary operations
+Consider an expression `e` of the form `e1 op e2`.
+
+#### Additive operators
+* `e1.unit` and `e2.unit` must be unit compatible.
+* `e.unit` must be compatible with `e1.unit`.
+* `e.unit` must be compatible with `e2.unit`.
+
+The three constraints ensure unit propagation even when one of the operand units is `empty` or `undefined`
+
+#### Multiplication
+`e.unit` is equal to `e1.unit * e2.unit`.
+
+#### Division
+`e.unit` is equal to `e1.unit * (e2.unit ^ -1)`.
+
+#### Relational operators
+The unit of `e1` and the unit of `e2` must be unit compatible.
+
+#### Power operator
+* If `e2` is a constant or evaluated Integer, `e.unit` is equal to `e1.unit ^ e2`. (Note that unit-checking does not trigger evaluation.)
+* Otherwise, `e1.unit` must be compatible with `"1"` and `e.unit` is equal to `"1"`.
+
+(If the exponent was not evaluated or is a Real, the unit of the base should be `"1"`.)
+
+(It is up for discussion, when `e2` is Real, whether `e2.unit` should be compatible with `"1"`.)
+
+### Function calls
+* For each expression passed as argument, its unit must be compatible with the declared unit of the corresponding function input.
+* If the output of the function has a declared unit, the unit of the expression is the declared unit.
+* The unit of the expression is `undefined` otherwise.
+
+### Operator der
+The unit of the expression `der(e)` is `der(e.unit)`.
+
+### If expressions
+Consider the expression `e` of the form `if cond then e1 else e2 end if;`
+* `e1.unit` and `e2.unit` must be unit compatible.
+* `e.unit` must be compatible with `e1.unit`.
+* `e.unit` must be compatible with `e2.unit`.
+
+(The presence of an `if`-expression during unit-checking means its condition was not evaluated.)
+
+### Equations
+Both sides of binding equations, equality equations and connect equations must be unit compatible.
+
+## Unit operators
+
+### `withUnit($value$, $unit$)`
+
+Create an expression with unit `$unit$`, and value `$value$`.
+Argument $value$ needs to be a `Real` expression, with empty unit.
+
+This operator makes it possible to attach a unit to a literal.
+
+### `withoutUnit($value$, $unit$)`
+Create an expression with empty unit, whose value is the numerical value of `$value$` expressed in `$unit$`.
+Argument `$value$` needs to be a `Real` expression, with a well-formed unit.
+The unit of `$value$` must be compatible with `$unit$`.
+
+### `inUnit($value$, $unit$)`
+Create an expression with unit `$unit$`, whose value is the conversion of `$value$` to `$unit$`.
+Argument `$value$` needs to be a `Real` expression, with a well-formed unit.
+The unit of `$value$` must be compatible with `$unit$`.
```

Comment:
No, not with the current definition of _compatible_.  Maybe the best way to fix this is to say _equivalent_ instead of _compatible_ for the unit constraints, and define `compatible` as just having the same base unit factors?

Reply:
Right, we should be able to convert from one to the other, so in other words they only differ by offset and scaling?